File: skills/code-optimizer/references/cpu/python/test_code.py
from collections.abc import Callable
import statistics
import time

from code import RecordItem, rank_bad, rank_good


def build_records(count: int) -> list[RecordItem]:
  records: list[RecordItem] = []
  for index in range(count):
    records.append(
      RecordItem(
        name=f"record-{index}-aeiou-{index % 17}",
        score=(index % 7) + 1,
      )
    )
  return records


def measure(run: Callable[[], int]) -> dict[str, float]:
  start_wall = time.perf_counter()
  start_cpu = time.process_time()
  run()
  elapsed_ms = (time.perf_counter() - start_wall) * 1000
  cpu_ms = (time.process_time() - start_cpu) * 1000
  return {
    "elapsed_ms": elapsed_ms,
    "cpu_ms": cpu_ms,
  }


def benchmark(label: str, run: Callable[[], int]) -> None:
  for _ in range(5):
    run()
  samples = [measure(run) for _ in range(25)]
  print(
    label,
    {
      "median_elapsed_ms": statistics.median(
        sample["elapsed_ms"] for sample in samples
      ),
      "median_cpu_ms": statistics.median(
        sample["cpu_ms"] for sample in samples
      ),
    },
  )


records = build_records(1500)
bad_result = rank_bad(records)
good_result = rank_good(records)

if bad_result != good_result:
  raise RuntimeError(f"Result mismatch: {bad_result} != {good_result}")

benchmark("rank_bad", lambda: rank_bad(records))
benchmark("rank_good", lambda: rank_good(records))

File: skills/code-optimizer/references/cpu/node/test_code.ts
import { rankBad, rankGood, type RecordItem } from "./code.ts";

type CpuUsage = {
  user: number;
  system: number;
};

declare const process: {
  cpuUsage(start?: CpuUsage): CpuUsage;
};

type Sample = {
  elapsedMs: number;
  userCpuMs: number;
  systemCpuMs: number;
};

function buildRecords(count: number): RecordItem[] {
  const records: RecordItem[] = [];
  for (let index = 0; index < count; index += 1) {
    records.push({
      name: `record-${index}-aeiou-${index % 17}`,
      score: (index % 7) + 1,
    });
  }
  return records;
}

function measure(run: () => number): Sample {
  const startCpu = process.cpuUsage();
  const start = performance.now();
  run();
  const elapsedMs = performance.now() - start;
  const cpu = process.cpuUsage(startCpu);
  return {
    elapsedMs,
    userCpuMs: cpu.user / 1000,
    systemCpuMs: cpu.system / 1000,
  };
}

function median(values: number[]): number {
  const ordered = [...values].sort((left, right) => left - right);
  const middle = Math.floor(ordered.length / 2);
  if (ordered.length % 2 === 0) {
    return (ordered[middle - 1] + ordered[middle]) / 2;
  }
  return ordered[middle];
}

function benchmark(label: string, run: () => number): void {
  for (let index = 0; index < 5; index += 1) {
    run();
  }
  const samples: Sample[] = [];
  for (let index = 0; index < 25; index += 1) {
    samples.push(measure(run));
  }
  console.log(label, {
    medianElapsedMs: median(samples.map((sample) => sample.elapsedMs)),
    medianUserCpuMs: median(samples.map((sample) => sample.userCpuMs)),
    medianSystemCpuMs: median(samples.map((sample) => sample.systemCpuMs)),
  });
}

const records = buildRecords(1500);
const badResult = rankBad(records);
const goodResult = rankGood(records);

if (badResult !== goodResult) {
  throw new Error(`Result mismatch: ${badResult} !== ${goodResult}`);
}

benchmark("rankBad", () => rankBad(records));
benchmark("rankGood", () => rankGood(records));
